generics/fanout_test.go

```diff
@@ -0,0 +1,196 @@
+package generics
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestFanout(t *testing.T) {
+	input := []int{1, 2, 3, 4, 5}
+	parallelism := 3
+	workerFactory := func(i int) (func(int) int, func(int)) {
+		worker := func(i int) int {
+			return i * 2
+		}
+		return worker, nil
+	}
+
+	result := Fanout(input, parallelism, workerFactory, nil)
+	assert.ElementsMatch(t, []int{2, 4, 6, 8, 10}, result)
+}
+
+func TestFanoutWithPredicate(t *testing.T) {
+	input := []int{1, 2, 3, 4, 5}
+	parallelism := 3
+	workerFactory := func(i int) (func(int) int, func(int)) {
+		worker := func(i int) int {
+			return i * 2
+		}
+		return worker, nil
+	}
+	predicate := func(i int) bool {
+		return i%4 == 0
+	}
+
+	result := Fanout(input, parallelism, workerFactory, predicate)
+	assert.ElementsMatch(t, []int{4, 8}, result)
+}
+
+func TestFanoutWithCleanup(t *testing.T) {
+	input := []int{1, 2, 3, 4, 5}
+	parallelism := 4
+	cleanupTotal := 0
+	mut := sync.Mutex{}
+	workerFactory := func(i int) (func(int) int, func(int)) {
+		worker := func(i int) int {
+			return i * 2
+		}
+		cleanup := func(i int) {
+			mut.Lock()
+			cleanupTotal += i
+			mut.Unlock()
+		}
+		return worker, cleanup
+	}
+
+	result := Fanout(input, parallelism, workerFactory, nil)
+	assert.ElementsMatch(t, []int{2, 4, 6, 8, 10}, result)
+	assert.Equal(t, 6, cleanupTotal) // 0 + 1 + 2 + 3
```

Comment:
Would it be easier to reason about to do `cleanupTotal += 1` and assert that `assert.Eqaul(t, parallelism, cleanupTotal)`?